Integrated CloudRift provider

Slonegg opened the pull request:
Add [CloudRift](https://www.cloudrift.ai/) as a provider. To start simple, I integrated a VM-based service and a few GPU types.

Provider requirements:
- [An API for creating and terminating VMs](https://api.cloudrift.ai/rapidoc#post-/api/v1/instances/rent)
- An external IP and a public port for SSH
  - Yes 
- Cloud-init
  - Yes. Here is an example of how to provide it for [https://api.cloudrift.ai/rapidoc#post-/api/v1/instances/rent](https://api.cloudrift.ai/rapidoc#post-/api/v1/instances/rent)
```
{
  "data": {
    "config": {
      "VirtualMachine": {
        "cloudinit_url": "https://storage.googleapis.com/cloudrift-vm-disks/cloudinit/ubuntu-base.cloudinit",
        "image_url": "https://storage.googleapis.com/cloudrift-vm-disks/disks/ubuntu/24.04/noble-server-cloudimg-amd64.img",
        "ssh_key": [
          "my-key"
        ]
      }
    },
    "recipe": "Ubuntu 24.04 Server",
    "selector": {
      "ByInstanceTypeAndLocation": {
        "datacenters": [
          "us-east-nc-nr-1"
        ],
        "instance_type": "generic-cpu.1"
      }
    },
    "with_public_ip": false
  },
  "version": "~upcoming"
}
```
- VM images with Ubuntu, OpenSSH, GPU drivers, and Docker with NVIDIA runtime
  - Yes, our default ones come with those above. If something is missing or needed in addition, we'll add it.

jvstme commented on src/tests/providers/test_cloudrift.py:
```diff
@@ -0,0 +1,20 @@
+
+def test_instance_types():
+    cloudrift_provider = CloudRiftProvider()
+    instance_types = cloudrift_provider.get()
```
This is a useful test, but we'd like to avoid real API calls in unit tests, as they can run often and should finish as soon as possible.

The best place for it depends on whether CloudRift will be an offline or an online provider (see the [contributing guide](https://github.com/dstackai/dstack/blob/master/contributing/BACKENDS.md), step 1.2)
- For offline providers, there are data integrity tests (see step 1.5)
- For online providers, we haven't set up integrity tests yet, but you can keep the test here and mock API calls with either pytest-mock (see [datacrunch](https://github.com/dstackai/gpuhunt/blob/main/src/tests/providers/test_datacrunch.py), [cudo](https://github.com/dstackai/gpuhunt/blob/main/src/tests/providers/test_cudo.py)) or requests-mock (see [vultr](https://github.com/dstackai/gpuhunt/blob/main/src/tests/providers/test_vultr.py))

Slonegg replied:
Modified the test to follow the example

jvstme commented on pyproject.toml:
```diff
@@ -55,7 +55,10 @@ oci = [
 ]
-all = ["gpuhunt[aws,azure,datacrunch,gcp,maybe_nebius,oci]"]
+cloudrift = [
+    "cloudrift>=0.46.3",
```
Is there a public repo with the source code for this library?

Slonegg replied:
Not at the moment. The Python client is part of our monorepo; we haven't yet extracted it.

Are you looking to make changes? The source code is generally available [here](https://pypi.org/project/cloudrift/#files). It is a pure Python library for making calls to https://api.cloudrift.ai/rapidoc

Slonegg commented:
> @Slonegg, thanks for the PR, it is indeed a good starting point. Before we can merge, can you please take a look at my review comments? And also these two:
> 
>     * Please register the provider in the catalog (see the [contributing guide](https://github.com/dstackai/dstack/blob/master/contributing/BACKENDS.md), step 1.4).
> 
>     * Please lint and format the code with `ruff check` and `ruff format` (I think this is not well documented for gpuhunt, we'll update the docs).
> 
> 
> Let me know if you'll need any assistance

Thanks! Apologies for not following the guide carefully.

> Please register the provider in the catalog (see the [contributing guide](https://github.com/dstackai/dstack/blob/master/contributing/BACKENDS.md), step 1.4).

>     Either OFFLINE_PROVIDERS or ONLINE_PROVIDERS in src/gpuhunt/_internal/catalog.py.
Added to OFFLINE_PROVIDERS

>     The python -m gpuhunt command in src/gpuhunt/__main__.py.
Modified the CLI and tested with with `python -m gpuhunt cloudrift --output cloudrift.txt

>     (offline providers) The CI workflow in .github/workflows/catalogs.yml.
Copied the runpod workflow and adjusted for cloudrift.

> Please lint and format the code with ruff check and ruff format (I think this is not well documented for gpuhunt, we'll update the docs).

Done

jvstme commented on pyproject.toml:
```diff
@@ -55,7 +55,7 @@ oci = [
     "datacrunch"
 ]
-all = ["gpuhunt[aws,azure,datacrunch,gcp,maybe_nebius,oci]"]
+all = ["gpuhunt[aws,azure,datacrunch,cloudrift,gcp,maybe_nebius,oci]"]
```
No longer relevant
```suggestion
all = ["gpuhunt[aws,azure,datacrunch,gcp,maybe_nebius,oci]"]
```

Slonegg replied:
Fixed

jvstme reviewed (approved):
@Slonegg, thanks for the fixes. There is one last review comment and a couple of linter errors in `ruff check` and `ruff format`. Hope you can fix these and I'm ready to merge

Slonegg commented:
> @Slonegg, thanks for the fixes. There is one last review comment and a couple of linter errors in `ruff check` and `ruff format`. Hope you can fix these and I'm ready to merge

Ah. I didn't realize that `ruff check` checks for more issues than what `ruff format` fixes. Addressed.

jvstme commented:
@Slonegg, I've just published gpuhunt 0.1.4 with CloudRift support. To continue the implementation in dstack, please update the gpuhunt version in [dstack/pyproject.toml](https://github.com/dstackai/dstack/blob/3a2689cf31b5e50773c920ff57e897b2c21e8bc7/pyproject.toml#L35) and reinstall the dependencies. Thanks for your contribution!
